### PR TITLE
Mandatory seq_profile and add marker bit.

### PR DESCRIPTION
**Pull request by tomfinegan**
Make the presence of seq_profile in AV1CodecConfigurationRecordv1
mandatory, and add marker bit to avoid producing config
records that share bit patterns similar to OBU headers.

### CODE REVIEW

**Comment by tomfinegan** [comment]
> Why not do "the very first bit of the AV1CodecConfigurationRecordv1 struct being always 1" instead of wasting a byte. Plus in the future we might not remember and start using the reserved byte ...

Fair point. How's it look with the marker bit?

File: index.bs [hunk 1]
```diff
@@ -254,6 +255,12 @@ Previous versions of this specification defined v0 of the AV1CodecConfigurationR
 
 The <dfn noexport>version</dfn> field indicates the version of the AV1CodecConfigurationRecordv1. The value SHALL be set to 1 for AV1CodecConfigurationRecordv1.
 
+The <dfn export>marker</dfn> field SHALL be set to 1.
+
+NOTE: The marker bit ensures that tje bit pattern of the first byte of the AV1CodecConfigurationRecordv1 cannot be mistaken for an [=OBU Header=] byte.
```
Comment (jamrial) [inline index.bs]: Typo, should be "the" not "tje". Same in index.html

Comment (tomfinegan) [reply index.bs]: Fixed. Thanks!

**Comment by jamrial** [comment]
>Fair point. How's it look with the marker bit?

seq_profile crosses the byte boundary (it has one bit in the first byte, the rest in the second). While is not a problem per se, I think it would be ideal if the first byte consisted on the marker and the version fields only, both of which are meant to never change location. So if any new version is defined in the future the only thing that's modified are the second byte onwards.

The change looks good otherwise. Thanks a lot for adding the marker.

**Comment by tomfinegan** [comment]
> seq_profile crosses the byte boundary (it has one bit in the first byte, the rest in the second). While is not a problem per se, I think it would be ideal if the first byte consisted on the marker and the version fields only, both of which are meant to never change location. So if any new version is defined in the future the only thing that's modified are the second byte onwards.

Added 7a09fc7 which changes the order. This puts marker, version, and seq_profile in the first byte while the reserved bits span from the end of the first byte into the second.

**Comment by porcelijn** [comment]
@cconcolato et al, I'm not sure I understand why the marker bit would be useful, as I cna't imaging how this part of the `av1C` would end up _inside_ an OBU stream. Also, more broadly, I don't understand why the av1-isobmff was changed to extend `av1C` from *Box* rather than *FullBox(0, 0)*.

It seems that the new set of properties (`seq_profile`, ... `chrome_sample_position`) would fit into the 24 bits regularly allotted to FullBox' `flags` to that the location of original `initial_presentation_delay` would not need to change if FullBox' `version=1`. The advantage of keeping it a FullBox is that it leaves the door open for future changes - all in the spirit of ISOBMFF.

To cite from ISO/IEC 14496-12:2015(E) appendix C.3:
> Unless absolutely unavoidable, boxes should contain either data (e.g. in fields), or other boxes, but not both. All boxes containing data should be a full box to allow later changes to syntax and semantics.  Boxes containing other boxes are known as container boxes, and are normally a plain (non‐full) box, since their semantics will never change if they are documented to contain only boxes.

I believe we should have something along these lines:
```
class AV1CodecConfigurationBox extends Box('av1C', 1, flags) {
  unsigned int (3) reserved = 0;
  unsigned int (1) initial_presentation_delay_present;
  if (initial_presentation_delay_present) {
    unsigned int (4) initial_presentation_delay_minus_one;
  } else {
    unsigned int (4) reserved = 0;
  }

  unsigned int (8)[] configOBUs;
}

Where `seq_profile = (flags & 0xC0000)`, ...
```

Or implicitly (as is done in the AVC spec., for example):
```
class AV1CodecConfigurationBox extends Box('av1C'){
  AV1CodecConfigurationRecord av1Config;
}

aligned (8) class AV1CodecConfigurationRecord {
  unsigned int (8) version = 1;
  unsigned int (3) seq_profile;
  unsigned int (5) seq_level_idx_0;
  unsigned int (1) seq_tier_0;
  unsigned int (1) high_bitdepth;
  unsigned int (1) twelve_bit;
  unsigned int (1) monochrome;
  unsigned int (1) chroma_subsampling_x;
  unsigned int (1) chroma_subsampling_y;
  unsigned int (2) chroma_sample_position;
  unsigned int (3) reserved = 0;

  unsigned int (1) initial_presentation_delay_present;
  if (initial_presentation_delay_present) {
    unsigned int (4) initial_presentation_delay_minus_one;
  } else {
    unsigned int (4) reserved = 0;
  }

  unsigned int (8)[] configOBUs;
}
```
The point being that ISOBMFF logic and assumptions (about which fields to expect in what version, future or backward compatibility, breaking change, etc.) would still apply.

Should I cross post this in https://github.com/AOMediaCodec/av1-isobmff/issues/71?

**Comment by jamrial** [comment]
>@cconcolato et al, I'm not sure I understand why the marker bit would be useful, as I cna't imaging how this part of the av1C would end up inside an OBU stream. Also, more broadly, I don't understand why the av1-isobmff was changed to extend av1C from Box rather than FullBox(0, 0).

So implementations propagating global headers can easily distinguish between raw obus or the contents of av1C. That bit is guaranteed to never be set in raw obus.

>Also, more broadly, I don't understand why the av1-isobmff was changed to extend av1C from Box rather than FullBox(0, 0).

So that the box version is also propagated.

IMO, the current state is ideal for all kinds of scenarios. And your second suggestion is the same thing currently in the spec, sans marker bit.

There really is no need to change anything.

**Comment by cconcolato** [comment]
> @cconcolato et al, I'm not sure I understand why the marker bit would be useful, as I cna't imaging how this part of the av1C would end up inside an OBU stream.

Some implementations initialize their decoders by passing the content of the config record and then passing OBUs. This allows distinguishing both.

> Also, more broadly, I don't understand why the av1-isobmff was changed to extend av1C from Box rather than FullBox(0, 0).

For the same reason as indicated above, we introduce a version in the config record, thus there was no need for duplicate versioning. Also avcC, hvcC are boxes not fullboxes. 

Also, we share that definition with Matroska: https://github.com/Matroska-Org/matroska-specification/blob/av1-mappin/codec/av1.md

> The advantage of keeping it a FullBox is that it leaves the door open for future changes

The version in the config record also allows that.
Anyway in general, versioning of the sample entry is tricky because it is not exposed at the 'codecs' parameter level for example. So a client cannot know before downloading the file, which version of the box it will receive. So, in the end changing the content in a non-backwards compatible way requires changing the 4CC. 

> all in the spirit of ISOBMFF.

I agree that in general FullBox should be used when containing fields and Box for containers, but that's not the practice for sample description entries (e.g avcC, hvcC).

> Or implicitly (as is done in the AVC spec., for example):

Interesting. You can already decide to implement it as a FullBox, the version would be 0b10000001.

**Comment by porcelijn** [comment]
(@jamrial)
> So implementations propagating global headers can easily distinguish between raw obus or the contents of av1C. That bit is guaranteed to never be set in raw obus.

I appreciate that this is technically a neat trick, but I think it's in the wrong place.

(@cconcolato)
> Some implementations initialize their decoders by passing the content of the config record and then passing OBUs.

Well - this spec isn't released yet, so I guess there's an opportunity to convince the implementers not to do that. In the end they will want to adjust their implementation to comply with the spec rather then the spec complying beta software.

I believe a FullBox is ideal for distinguishing between different variants - you might designate bits in flags to indicate whether it holds a config record, OBUs (or both).

> Also avcC, hvcC are boxes not fullboxes.

That is technically true, but at least the first byte behaves the way a FullBox version would. In comparison, consider VP9's `vpcC` as an example of where a FullBox *is* used:
```
class VPCodecConfigurationBox extends FullBox('vpcC', version = 1, 0){
      VPCodecConfigurationRecord() vpcConfig;
}

aligned (8) class VPCodecConfigurationRecord {
    unsigned int (8)     profile;
    unsigned int (8)     level;
    unsigned int (4)     bitDepth;
    unsigned int (3)     chromaSubsampling;
    unsigned int (1)     videoFullRangeFlag;
    unsigned int (8)     colourPrimaries;
    unsigned int (8)     transferCharacteristics;
    unsigned int (8)     matrixCoefficients;
    unsigned int (16)    codecIntializationDataSize;
    unsigned int (8)[]   codecIntializationData;
}
```
> Also, we share that definition with Matroska: https://github.com/Matroska-Org/matroska-specification/blob/av1-mappin/codec/av1.md

Could be mistaken about matroska's spec, but I think that is in motion as well. In fact, they seem to be looking at us for answers on this topic https://www.ietf.org/mail-archive/web/cellar/current/msg01650.html

> Anyway in general, versioning of the sample entry is tricky because it is not exposed at the 'codecs' parameter level for example. So a client cannot know before downloading the file, which version of the box it will receive. So, in the end changing the content in a non-backwards compatible way requires changing the 4CC.

That's an interesting point, regardless of FullBox or as in the current spec. I guess that unless you decide to add version to codec string, you indeed end up with an `av1C` whose version is effectively either frozen ("MUST be 1") or irrelevant ("don't care") but in either case cannot be used to introduce new field to a future version of the box. I think this is precisely demonstrates why 14496-12 recommends using FullBox for fields.

> You can already decide to implement it as a FullBox, the version would be 0b10000001.

Sure, but it smells weird to use version=129 before the first release of the av1-isobmff spec is out the door, doesn't it? :smirk:

I have a new suggestion. Why don't we keep the marker bit in AV1CodecConfigurationRecord but drop version from this record and instead use a FullBox for AV1CodecConfigurationBox:
```
class AV1CodecConfigurationBox extends FullBox('av1C', 1, 0){
  AV1CodecConfigurationRecord av1Config;
}

aligned (8) class AV1CodecConfigurationRecord {
  unsigned int (1) marker = 1;
  unsigned int (3) seq_profile;
  unsigned int (5) seq_level_idx_0;
  unsigned int (1) seq_tier_0;
  unsigned int (1) high_bitdepth;
  unsigned int (1) twelve_bit;
  unsigned int (1) monochrome;
  unsigned int (1) chroma_subsampling_x;
  unsigned int (1) chroma_subsampling_y;
  unsigned int (2) chroma_sample_position;
  unsigned int (3) reserved = 0;

  unsigned int (1) initial_presentation_delay_present;
  if (initial_presentation_delay_present) {
    unsigned int (4) initial_presentation_delay_minus_one;
  } else {
    unsigned int (4) reserved = 0;
  }

  unsigned int (8)[] configOBUs;
}
```
(Okay, this screws up the `AV1CodecConfigurationRecord` bit alignment so that would need to be somewhat rethought, obviously...)

**Comment by jamrial** [comment]
>That is technically true, but at least the first byte behaves the way a FullBox version would. In comparison, consider VP9's vpcC as an example of where a FullBox is used:

vpcC does not require the propagation of global headers whatsoever. It would have if vp10 ended up existing, but since it didn't, the new aom specific box was devised using proven ideas from avcC and hvcC for this purpose.

>I have a new suggestion. Why don't we keep the marker bit in AV1CodecConfigurationRecord but drop version from this record and instead use a FullBox for AV1CodecConfigurationBox:

Because then the version is not propagated. As mentioned earlier, that's the reason avcC and hvcC both use Box with a custom version field, instead of a FullBox.

**Comment by cconcolato** [comment]
@porcelijn I am not in favor of the changes you propose. The current spec is not wrong and I don't see the benefit of your proposal. I will expose your comment during today's AOM call and we'll decide.

**Comment by porcelijn** [comment]
Okay. Maybe this is not such a big deal as I have been making out of this and there seems to be some consensus between you guys to keep it as is. :relaxed:
Thanks for considering my comments!

**Pull request merged**